### PR TITLE
Adds Pulumi AI integrations with Pulumi New

### PR DESCRIPTION
**Pull request by kpitzen**
# Description

Now that we support .zip archive sources for Pulumi New, we have all of the API surface we need to provide a full Pulumi New experience using Pulumi AI. This PR introduces a few modes of interacting with Pulumi AI to generate Pulumi projects.

- The default `pulumi new` experience now begins with a choice between `ai` and `template` modes - the `template` mode represents the current state of running `pulumi new`, while `ai` provides an interactive experience with Pulumi AI.
  - The user can iteratively ask questions to improve or change the resulting program - each time a prompt is completed, they are asked to `refine`, `no`, or `yes` their session - `refine` allows a follow-on prompt to be submitted. `no` ends the session without generating a pulumi project, and `yes` generates a Pulumi project from the most recent program returned by Pulumi AI.
- Additionally, top-level flags, `--ai` and `--language` are provided to fill in default values for the AI mode. When a prompt is provided with a language, it is automatically submitted to Pulumi AI - if either is missing, the user is prompted for whichever value is missing.

Fixes https://github.com/pulumi/pulumi.ai/issues/441
Fixes https://github.com/pulumi/pulumi.ai/issues/443
Fixes https://github.com/pulumi/pulumi.ai/issues/444

Depends on https://github.com/pulumi/pulumi.ai/pull/472
Depends on https://github.com/pulumi/pulumi.ai/pull/507
Depends on https://github.com/pulumi/pulumi.ai/pull/508

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/cmd/pulumi/new.go [hunk 1]
```diff
@@ -390,6 +420,195 @@ func useSpecifiedDir(dir string) (string, error) {
...
+		name,
+		"",
+	); continuePrompt == "continue"; continuePrompt, conversationURL, _, err = runAINewPromptStep(
+		opts,
+		language,
+		prompt,
+		yes,
+		name,
+		continuePrompt,
+	) {
```
Comment (AaronFriel) [inline pkg/cmd/pulumi/new.go:491]: I think there's a bug here where the connection ID isn't retrieved and passed to the subsequent calls. 

This `for expr; check; expr; { }` loop is a bit strange to me, too, hard to parse. Is this essentially a "do while"?

**Review by AaronFriel — COMMENTED** [review]
A couple comments. 

When running, I couldn't get the continuation to update the existing program, it looks like it didn't pass the conversation & connect ID through to subsequent calls.

The "continue/done/abort" prompt confused me at first - think we could probably reframe the question as "Use this program as a template?" "Yes/no/[clarify/refine/etc]"?

File: pkg/cmd/pulumi/new.go (hunk 1 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:491]: > This for expr; check; expr; { } loop is a bit strange to me, too, hard to parse. Is this essentially a "do while"?

That's correct - there aren't many ergonomic ways to express that kind of control flow in golang, and this was about as clean as I could find.

It's definitely possible the connectionID isn't being passed through correctly, so let me take a look (though I do believe I tested this - never hurts to verify).

Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:491]: Ah - I see the issue re: connectionID. Will fix this now.

**Review by AaronFriel — APPROVED** [review]
LGTM, opening the thread in Platform for a release later today or Monday.

File: pkg/cmd/pulumi/new.go [hunk 2]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+		ResponseMode:   "code",
+		ConversationID: conversationID,
+		ConnectionID:   connectionID,
+	}
+	marshalledBody, err := json.Marshal(requestBody)
+	if err != nil {
+		return "", "", "", err
+	}
+	userDataCookie := http.Cookie{Name: "pulumi_command_line_user_name", Value: userName}
+	request, err := http.NewRequest("POST", requestPath.String(), bytes.NewReader(marshalledBody))
```
Comment (justinvp) [inline pkg/cmd/pulumi/new.go:550]: Interesting. So you don't actually have to be logged in to Pulumi Cloud to make these requests?

File: pkg/cmd/pulumi/new.go [hunk 3]
```diff
@@ -142,6 +165,44 @@ func runNew(ctx context.Context, args newArgs) error {
...
+		if shouldPromptForAIOrTemplate(args, b) {
+			aiOrTemplate, err = chooseWithAIOrTemplate(opts)
+		} else {
+			aiOrTemplate = deriveAIOrTemplate(args)
+		}
+		if err != nil {
+			return err
+		}
+		if aiOrTemplate == "ai" {
+			if b.Name() != "pulumi.com" {
```
Comment (justinvp) [inline pkg/cmd/pulumi/new.go:194]: Do we want AI prompts to work for other instances of Pulumi Cloud, e.g. staging, dev stacks, etc? Or only the production service?

File: pkg/cmd/pulumi/new.go [hunk 4]
```diff
@@ -142,6 +165,44 @@ func runNew(ctx context.Context, args newArgs) error {
...
+				// Currently, we don't actually hit this - users are prompted to log in
+				// for pulumi new regardless, but if that behavior changes, we can catch it here.
+				return errors.New("please log in to Pulumi Cloud to use Pulumi AI")
```
Comment (justinvp) [inline pkg/cmd/pulumi/new.go:197]: Note: We do hit this if you're logged-in to the filestate backend (or I guess if you're logged into some other non-Pulumi Cloud URL), e.g.:

```
[~/Documents/2023/12/11/newtesttest]$ pulumi login --local
Logged in to Justins-Pulumi-MacBook-Pro.local as justin (file://~)
[~/Documents/2023/12/11/newtesttest]$ pulumi new --ai "foo"
error: please log in to Pulumi Cloud to use Pulumi AI
```

So the comment isn't really accurate.

File: pkg/cmd/pulumi/new.go [hunk 5]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
 	return cwd, nil
 }
 
+type AIPromptRequestBody struct {
```
Comment (justinvp) [inline pkg/cmd/pulumi/new.go:454]: Nit: I do wonder if we should factor some of the AI-specific utilities out of `new.go` to a separate file.

Also, does this struct need to be exported?

**Review by justinvp — APPROVED** [review]
The changes generally look good. I do think there's room for refinement, but that could be done in subsequent changes if we want to get this landed sooner than later.

## Bug 1

I found one bug. If I run `pulumi new --ai "an aws bucket" --language foo` with an invalid language, I get:

Screen Shot 2023-12-12 at 5 19 17 AM

Which can't be addressed by any of those options.

## Bug 2 (maybe?)

Perhaps this isn't a bug, but if I run `pulumi new --ai --language typescript`

It immediately asks me for project name, description, and stack name... Basically, it creates an empty project.

Screen Shot 2023-12-12 at 5 27 52 AM

Which may indeed be the desired behavior. But I am wondering if `--ai` is specified and the value is empty and we're interactive, if it should prompt the user to fill in an actual prompt.

## Colors

There's inconsistent color use in the prompts. Oddly, if I start with just `pulumi new`, I don't get any colored text at all in the prompts:

Screen Shot 2023-12-11 at 6 46 23 PM

Screen Shot 2023-12-11 at 6 46 32 PM

Screen Shot 2023-12-11 at 6 46 51 PM

However, if I start with `pulumi new --ai "an aws bucket"`, I do get colors in prompts:

Screen Shot 2023-12-11 at 6 45 48 PM

Screen Shot 2023-12-11 at 6 46 07 PM

But note: The colors used aren't the typical colors we use for prompts. Generally, the prompt message should be light blue:

Screen Shot 2023-12-12 at 5 31 59 AM

## Pulumi AI prompt text

```
Please input your prompt here:

```

We're sort of assuming folks are familiar with Pulumi AI and prompts. Will users know what they should enter here? (I suppose you'd have to be living under a rock). Still, I wonder if we should help users out with some more text and a canonical example (I show "An AWS bucket" below, but we may have a better example). Something like:

```
Please input a prompt for Pulumi AI and press <ENTER>.
For example: "An AWS bucket"
Pulumi AI prompt:

```

Screen Shot 2023-12-12 at 5 53 41 AM

## Tests

Where are the tests? Would be good if the prompts and API requests/responses were mockable so we could write tests like the tests in `new_test.go`.

File: pkg/cmd/pulumi/new.go [hunk 6]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+		connectionID,
+		userName,
+		language,
+	)
+	if err != nil {
+		return "", "", "", "", err
+	}
+	if connectionID != "" && connectionID != connectionIDReturn {
+		connectionIDReturn = connectionID
+		fmt.Println("Connection ID changed, please restart the prompt")
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:619]: What does this mean - and is it intentional to print it here?

File: pkg/cmd/pulumi/new.go [hunk 7]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+	}
+	parsedURL, err := url.Parse(pulumiAIURL)
+	if err != nil {
+		return "", "", "", err
+	}
+	requestPath := parsedURL.JoinPath("api", "chat")
+	requestBody := AIPromptRequestBody{
+		Language:       language,
+		Instructions:   promptMessage,
+		Model:          "gpt-4-turbo",
```
[the remark this answers is not in the document]
Comment (lukehoban) [reply pkg/cmd/pulumi/new.go:540]: I don't think we should.  Can the service maintain a default for this that we trust from this interface?  And then we can update the service as we want?

File: pkg/cmd/pulumi/new.go [hunk 8]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+	}
+	request.AddCookie(&userDataCookie)
+	fmt.Println("Sending prompt to Pulumi AI...")
+	res, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return "", "", "", err
+	}
+	defer res.Body.Close()
+	reader := bufio.NewReader(res.Body)
+	fmt.Println("Pulumi AI response:")
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:562]: I'm seeing the output from this formatted a bit differently than the rest of the output - the leading space gets "undone".  Maybe intentional?

```
 Would you like to create a new project with Pulumi AI? ai
 Please select a language for your project: Go
 Please input your prompt here:
 NGINX deployed to kubernetes deployment and service with public IP address
Sending prompt to Pulumi AI...
Pulumi AI response:
```

File: pkg/cmd/pulumi/new.go [hunk 9]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+		if err != nil && err.Error() == "EOF" {
+			// Add a newline to ensure we don't overwrite the last line of the Pulumi AI response
+			fmt.Println()
+			break
+		}
+	}
+	conversationID = res.Header.Get("x-conversation-id")
+	connectionID = res.Header.Get("x-connection-id")
+	conversationURL := parsedURL.JoinPath("api", "project", url.PathEscape(fmt.Sprintf("%s.zip", conversationID))).String()
+	return conversationURL, connectionID, conversationID, nil
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:578]: Suggest we print out the URL to the question in pulunmi.com/ai at the end here - so that the user can navigate there and have a permanent record of the question and answer used for the template is available.  Also connects this more explicitly to the rest of Pulumi AI.

File: pkg/cmd/pulumi/new.go [hunk 10]
```diff
@@ -843,6 +1114,35 @@ func pythonCommands() []string {
...
+	// Customize the prompt a little bit (and disable color since it doesn't match our scheme).
+	surveycore.DisableColor = true
+
+	options := []string{
+		"ai",
+		"template",
+	}
+
+	optionsDescriptionMap := map[string]string{
+		"ai":       "Create a new Pulumi project using Pulumi AI",
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:1128]: I expect we should default to the normal template for now, instead of defaulting to the AI option.  As a result I also think we should invert the phrasing of the prompt (or make it more neutral) like "Would you like to create a project from a template or using a Pulumi AI prompt?".

I still do wonder if there is some way for the initial prompt to go straight into the template list if you type anything (like `a` for `aws-typescript`), but stay in the AI vs. template selector if you press up/down/enter.

File: pkg/cmd/pulumi/new.go [hunk 11]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+	args newArgs,
+	opts display.Options,
+	userName string,
+) (conversationURL string, err error) {
+	languageOptions := []string{
+		"TypeScript",
+		"JavaScript",
+		"Python",
+		"Go",
+		"C#",
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:476]: Is there a reason we don't offer YAML and Java options here?  I think we should (even if they are less used).

File: pkg/cmd/pulumi/new.go [hunk 12]
```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
...
+) (continueSelection string,
+	conversationURLReturn string,
+	conversationIDReturn string,
+	connectionIDReturn string,
+	err error,
+) {
+	var promptMessage string
+	if prompt == "" || currentContinueSelection != "" {
+		if err := survey.AskOne(&survey.Input{
+			Message: "Please input your prompt here:\n",
```
Comment (lukehoban) [inline pkg/cmd/pulumi/new.go:600]: Perhaps offer a suggestion like in pulumi.com/ai to avoid "blank screen" problems?

```suggestion
			Message: "Please input your prompt here ("a static website on AWS behind a CDN"):\n",
```

File: pkg/cmd/pulumi/new.go (hunk 5 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:454]: I wondered the same thing here - I'm perfectly happy to pull all the AI stuff into a separate file here, but wasn't sure what the desired pattern was. I'll get that stuff moved now (it doesn't need to be exported, no).

File: pkg/cmd/pulumi/new.go (hunk 4 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:197]: Correct! I'd meant to remove this after testing (I hadn't considered filestate backends, so it makes sense to remove this comment).

File: changelog/pending/20231207--cli-new--adds-pulumi-new-ai-functionality.yaml [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  description: |
+  Adds pulumi new --ai functionality.
+  The `pulumi new` command now defaults to a new experience - users can select `ai` or `template` if no arguments are specified.
+
+  If `ai` is selected, we offer a guided experience to iteratively prompt Pulumi AI to craft a program for you.
+
+  Additionally, for a less interactive experience, we offer the flags `--ai` and `--language`. If `--ai <prompt>` is provided,
+  we will begin the interactive prompting session with the provided prompt. If `--language <language>` is provided with a supported language,
+  we will skip the language selection prompt and begin the interactive prompting session with the provided language. Addionally, if `--yes` is provided,
+  we will skip the verification step and immediately create the project generated by the given prompt and language.
```
Comment (lukehoban) [inline changelog/pending/20231207--cli-new--adds-pulumi-new-ai-functionality.yaml:13]: Would be great to add some of this same information into the docs for `pulumi new` as well (in pkg/cmd/pulumi/new.go).

File: pkg/cmd/pulumi/new.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (AaronFriel) [reply pkg/cmd/pulumi/new.go:540]: @kpitzen we should probably move this into the AI endpoint as a default, if it isn't.

File: pkg/cmd/pulumi/new.go (hunk 3 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:194]: That's a great question - I _think_, because this feature doesn't actually interact with p-cloud (all the integrations are with the AI service directly) that leaving it as the production service makes sense. I'd be more open to configuring this / being a bit more clever here if we wanted to test upstream integrations with other backends, but that's not really relevant here.

File: pkg/cmd/pulumi/new.go (hunk 2 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:550]: You do not - there are some details about auth impl on the AI service that don't require it (specifically, we allow one prompt for unauthenticated users).

File: pkg/cmd/pulumi/new.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:540]: I wouldn't mind pulling this into an argument with a default - would that be your preference as well?

**Review by lukehoban — COMMENTED** [review]
Overall looking and working nicely!

One comment that I think is related to the backend.  When I generated a Go program, it generated this invalid go.mod:
```
module ainew
  
  go 1.21

  toolchain go1.21.1

  require (
github.com/pulumi/pulumi/sdk/v3/go/pulumi v3
)
```

which errors with:

```
go: errors parsing go.mod:
go.mod:8: no matching versions for query "v3"
```

File: pkg/cmd/pulumi/new.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go:540]: I've opened this PR to resolve this issue:
https://github.com/pulumi/pulumi.ai/pull/507

[the remark this answers is not in the document]
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: Also updated this PR to reflect the above.

File: pkg/cmd/pulumi/new.go (hunk 6 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: I've yet to reliably recreate an issue I saw earlier on in development - it can be the case where a "new session" is started if a request is missing user information, or if user information changes from one prompt to the next. This may leak some of the backend implementation, but when that happens, we generate a new "connection" object, which is what is used to determine how many follow-up prompts a user has made in the same conversation. It's also the metric we use to determine whether to prompt a user to log in to p-cloud after a number of unauthenticated prompts.

File: pkg/cmd/pulumi/new.go (hunk 10 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: Interesting - the feedback we got during the design phase was to default to AI (put it on top). I can certainly reverse the order, though.

File: pkg/cmd/pulumi/new.go (hunk 11 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: We didn't have YAML because it wasn't yet supported in the AI service for project rendering. I've added that support here:
https://github.com/pulumi/pulumi.ai/pull/508
and have updated this PR to support both. (Java was just a simple omission)

File: pkg/cmd/pulumi/new.go (hunk 8 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: I suspect that's related to `Survey` formatting vs just standard `println` - I'm happy to reformat it in whatever way makes the most sense here.

**Comment by kpitzen** [comment]
@justinvp 
> Perhaps this isn't a bug, but if I run pulumi new --ai --language typescript

> It immediately asks me for project name, description, and stack name... Basically, it creates an empty project.

This seems to be a Cobra issue - when running that way, this is how args get parsed:
image

I'm not sure how to check for "the prompt message is also the name of a flag" in this case.

File: pkg/cmd/pulumi/new.go (hunk 10 above)
Comment (kpitzen) [reply pkg/cmd/pulumi/new.go]: > I still do wonder if there is some way for the initial prompt to go straight into the template list if you type anything (like a for aws-typescript), but stay in the AI vs. template selector if you press up/down/enter.

I did look into this as I was starting work here - it seems to me that it's impossible when using `Survey`. I would _love_ to be wrong, though.

**Comment by kpitzen** [comment]
@justinvp - I've resolved the coloring inconsistency wrt `pulumi new` vs `pulumi new --some-flags`. I think a reasonable follow-up to this would be to implement our current colorization patterns to these prompts (as `pulumi new` formerly just disabled colors entirely).

File: pkg/cmd/pulumi/new_ai_test.go
Comment (justinvp) [inline pkg/cmd/pulumi/new_ai_test.go:1]: Nit: copyright header

File: pkg/cmd/pulumi/new_ai.go [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+			// Add a newline to ensure we don't overwrite the last line of the Pulumi AI response
+			fmt.Println()
+			break
+		}
+	}
+	conversationID = res.Header.Get("x-conversation-id")
+	connectionID = res.Header.Get("x-connection-id")
+	projectURL := parsedURL.JoinPath("api", "project", url.PathEscape(fmt.Sprintf("%s.zip", conversationID))).String()
+	conversationURL := parsedURL.JoinPath("conversations", conversationID).String()
+	fmt.Println("View this conversation at:", conversationURL)
```
Comment (justinvp) [inline pkg/cmd/pulumi/new_ai.go:163]: Should there be a space after the colon?

```suggestion
	fmt.Printf("View this conversation at: %s\n", conversationURL)
```

File: pkg/cmd/pulumi/new.go [hunk 15]
```diff
@@ -535,6 +578,17 @@ func newNewCmd() *cobra.Command {
 	cmd.PersistentFlags().BoolVarP(
 		&args.listTemplates, "list-templates", "l", false,
 		"List locally installed templates and exit")
+	cmd.PersistentFlags().StringVar(
+		&args.aiPrompt, "ai", "", "Prompt to use for Pulumi AI",
+	)
+	cmd.PersistentFlags().Var(
+		&args.aiLanguage, "language", "Language to use for Pulumi AI "+
+			"(possible choices: TypeScript, JavaScript, Python, Go, C#)",
```
Comment (justinvp) [inline pkg/cmd/pulumi/new.go:586]: Need to update this for YAML and Java.

File: pkg/cmd/pulumi/new_ai.go [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+	language httpstate.PulumiAILanguage,
+) (string, string, string, error) {
+	pulumiAIURL := env.AIServiceEndpoint.Value()
+	if pulumiAIURL == "" {
+		pulumiAIURL = "https://www.pulumi.com/ai"
+	}
+	parsedURL, err := url.Parse(pulumiAIURL)
+	if err != nil {
+		return "", "", "", err
+	}
```
Comment (justinvp) [inline pkg/cmd/pulumi/new_ai.go:131]: Should the URL somehow be encapsulated within the backend? Is it always going to be `https://www.pulumi.com/ai`, even with self-hosted backends?

**Review by justinvp — APPROVED** [review]
Some comments, otherwise LGTM. (I haven't actually tried these latest changes yet. I've only looked over the code).

File: pkg/cmd/pulumi/new.go (hunk 15 above)
Comment (AaronFriel) [reply pkg/cmd/pulumi/new.go:586]: We've omitted these for now as the code quality is not at the level we feel comfortable with.

File: pkg/cmd/pulumi/new_ai.go (hunk 16 above)
Comment (AaronFriel) [reply pkg/cmd/pulumi/new_ai.go:131]: Yes - the zip URLs are located there presently. We don't expect self-hosted to be supported for a while yet after this release as we need to address other complexity around hosting Pulumi AI.

File: pkg/cmd/pulumi/new.go (hunk 15 above)
Comment (AaronFriel) [reply pkg/cmd/pulumi/new.go:586]: On reflection, I'll update the PR to add these languages and let issues filed so that we get user requests to drive specific improvement here.

**Review by justinvp — APPROVED** [review]
